src/pykiso/lib/connectors/trc_handler.py

```diff
@@ -0,0 +1,322 @@
+        msg.is_error_frame = type_ == "ER"
+        msg.is_remote_frame = type_ == "RR"
+
+        return TypedMessage(type_, msg)
+
+
+class TRCWriterCanFD(TRCWriter):
+    """Logs CAN FD data to text file (.trc)"""
+
+    # Type has been added to FORMAT_MESSAGE >= 2.0
+    FORMAT_MESSAGE_V2_1 = "{msgnr:>7} {time:13.3f} {type:>2} {channel:>2} {id:>8} {dir:>2} - {dlc:<4} {data}"
+
+    FORMAT_MESSAGE_V2_0 = (
+        "{msgnr:>7} {time:13.3f} {type:>2} {id:>8} {dir:>2} {dlc:<2} {data}"
+    )
+
+    def _format_message_init(self, msg: TypedMessage, channel: int) -> str:
+        """Pick message format from file version and format initial message
+
+        :return: message informations formated in a string
+        """
+        # Fix error from python can -> message number should start at one not zero
+        self.msgnr = 1
+
+        if self.file_version == TRCFileVersion.V1_0:
+            self._msg_fmt_string = self.FORMAT_MESSAGE_V1_0
+        elif self.file_version == TRCFileVersion.V2_0:
+            self._msg_fmt_string = self.FORMAT_MESSAGE_V2_0
+        elif self.file_version == TRCFileVersion.V2_1:
+            self._msg_fmt_string = self.FORMAT_MESSAGE_V2_1
+        else:
+            raise NotImplementedError("File format is not supported")
+
+        self._format_message = self._format_message_by_format
+        return self._format_message_by_format(msg, channel)
+        
+    def _format_message_by_format(self, msg: TypedMessage, channel: int):
+        """Format messages
+
+        :return: message informations formated in a string
+        """
+        if msg.type not in ["EC", "ER", "ST"]:
+            if msg.is_extended_id:
+                arb_id = f"{msg.arbitration_id:07X}"
+            else:
+                arb_id = f"{msg.arbitration_id:04X}"
+        else:
+            arb_id = f"{msg.arbitration_id}"
+        data = [f"{byte:02X}" for byte in msg.data]
+
+        #For the time python-can was doing substraction with the first message timestamp
+        #which is incorrect. It should be with the start time of the trace otherwise the 
+        #first message will always have an offset of zero.
+        if self.file_version == TRCFileVersion.V1_0:
+            serialized = self._msg_fmt_string.format(
+                msgnr=self.msgnr,
+                time=(msg.timestamp - self.first_timestamp) * 1000,
+                channel=channel,
+                id=arb_id,
+                dir="Rx" if msg.is_rx else "Tx",
+                dlc=msg.dlc,
+                data=" ".join(data),
+            )
+        else:
+            serialized = self._msg_fmt_string.format(
+                msgnr=self.msgnr,
+                time=(msg.timestamp - self.first_timestamp) * 1000,
+                type=msg.type,
+                channel=channel,
+                id=arb_id,
+                dir="Rx" if msg.is_rx else "Tx",
+                dlc=msg.dlc,
+                data=" ".join(data),
+            )
```

Comment:
A lot of the parameters seem to be repeated, maybe you can do something like:
`
format_fields = {
        "msgnr": self.msgnr,
        "time": (msg.timestamp - self.first_timestamp) * 1000,
        "channel": channel,
        "id": arb_id,
        "dir": "Rx" if msg.is_rx else "Tx",
        "dlc": msg.dlc,
        "data": " ".join(data),
}
if self.file_version != TRCFileVersion.V1_0:
    format_fields["type"] = (msg.timestamp - self.first_timestamp) * 1000

return self._msg_fmt_string.format(**format_fields)
`

Reply:
Good suggestion, but I think we'll focus on readability and leave it as it is it that's fine for you